# KNN model

## Loading data

Step before loading date: Using excel first combined all the data in one sheet.

In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('/Users/yunlei/Desktop/MGMT 478/Combined dataset_nonsort.csv')
data.head(20)

STATION                                        NAME  LATITUDE  \
0   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
1   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
2   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
3   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
4   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
5   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
6   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
7   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
8   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
9   USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
10  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
11  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
12  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
13  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
14  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
15  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
16  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
17  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
18  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   
19  USW00014835  LAFAYETTE PURDUE UNIVERSITY AIRPORT, IN US  40.41236   

    LONGITUDE  ELEVATION     DATE  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  
0   -86.94739      181.7  2010-01   8.5  1.01   NaN  23.5  29.5  17.5  
1   -86.94739      181.7  2010-02   7.6  0.61   NaN    26  32.8  19.1  
2   -86.94739      181.7  2010-03   7.2  3.22   NaN  44.8  55.1  34.6  
3   -86.94739      181.7  2010-04   8.1  2.49   NaN  58.1  70.4  45.8  
4   -86.94739      181.7  2010-05   6.7  5.55   NaN  64.7  75.2  54.2  
5   -86.94739      181.7  2010-06   5.6  9.61   NaN  74.6  84.3  64.8  
6   -86.94739      181.7  2010-07   4.3  4.04   NaN  76.8  86.8  66.9  
7   -86.94739      181.7  2010-08   3.8  2.23   NaN  76.8  87.7    66  
8   -86.94739      181.7  2010-09   6.0  1.81   NaN  67.5    80    55  
9   -86.94739      181.7  2010-10   6.3  0.97   NaN  56.4  70.4  42.4  
10  -86.94739      181.7  2010-11   7.4  3.50   NaN  43.6  55.2    32  
11  -86.94739      181.7  2010-12   7.4  1.26   NaN  24.1    31  17.2  
12  -86.94739      181.7  2011-01   7.6  0.73   NaN  22.7  30.1  15.3  
13  -86.94739      181.7  2011-02   8.5  2.66   NaN  31.3  39.6    23  
14  -86.94739      181.7  2011-03   9.2  2.15   NaN  42.3  52.1  32.5  
15  -86.94739      181.7  2011-04  10.1  7.89   NaN  54.5  65.4  43.7  
16  -86.94739      181.7  2011-05   6.7  1.32   NaN  63.5  73.4  53.7  
17  -86.94739      181.7  2011-06   6.3  6.51   NaN    73  82.7  63.4  
18  -86.94739      181.7  2011-07   4.3  3.41   NaN  79.9    90  69.8  
19  -86.94739      181.7  2011-08   4.3  4.97   NaN  74.4  85.6  63.1

In [3]:
# Check for missing values
data.isnull().sum()

STATION         0
NAME            0
LATITUDE        0
LONGITUDE       0
ELEVATION       0
DATE            0
AWND          700
PRCP            8
SNOW         1713
TAVG           27
TMAX           10
TMIN           25
dtype: int64

## Cleaning data

In [4]:
from sklearn.impute import SimpleImputer

In [5]:
data['DATE'] = pd.to_datetime(data['DATE'])
data.sort_values(['STATION', 'DATE'], inplace=True)

In [6]:
data.set_index(['STATION', 'DATE'], inplace=True)

In [7]:
for lag in [1, 3, 7]:
    data[f'PRCP_Lag{lag}'] = data.groupby(level='STATION')['PRCP'].shift(lag)

In [8]:
data.head(20)

NAME  LATITUDE  LONGITUDE  \
STATION     DATE                                                           
USC00114355 2010-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-07-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   

                        ELEVATION  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  \
STATION     DATE                                                        
USC00114355 2010-01-01      167.6   NaN  2.00   7.6   NaN  23.9   NaN   
            2010-02-01      167.6   NaN  0.80  12.8   NaN  31.1   NaN   
            2010-03-01      167.6   NaN  2.50   1.4   NaN  52.8   NaN   
            2010-04-01      167.6   NaN  3.34   0.0   NaN  71.4   NaN   
            2010-05-01      167.6   NaN  6.63   0.0   NaN    73   NaN   
            2010-06-01      167.6   NaN  9.96   0.0   NaN  83.6   NaN   
            2010-08-01      167.6   NaN  5.50   0.0   NaN    86   NaN   
            2010-09-01      167.6   NaN  4.76   0.0   NaN  76.6   NaN   
            2010-10-01      167.6   NaN  1.67   0.0   NaN  69.9   NaN   
            2010-11-01      167.6   NaN  1.55   0.0   NaN  52.8   NaN   
            2010-12-01      167.6   NaN  1.55  11.0   NaN  30.4   NaN   
            2011-01-01      167.6   NaN  0.89   NaN   NaN  26.6   NaN   
            2011-02-01      167.6   NaN  2.67   NaN   NaN  35.4   NaN   
            2011-03-01      167.6   NaN  2.07   NaN   NaN  48.4   NaN   
            2011-04-01      167.6   NaN  3.64   NaN   NaN  58.7   NaN   
            2011-05-01      167.6   NaN  3.51   NaN   NaN  73.4   NaN   
            2011-06-01      167.6   NaN  4.71   NaN  71.7  80.9  62.6   
            2011-07-01      167.6   NaN  2.02   0.0  80.6  89.8  71.4   
            2011-08-01      167.6   NaN  0.81   NaN  75.1  85.1    65   
            2011-09-01      167.6   NaN  2.21   NaN  62.3  70.9  53.7   

                        PRCP_Lag1  PRCP_Lag3  PRCP_Lag7  
STATION     DATE                                         
USC00114355 2010-01-01        NaN        NaN        NaN  
            2010-02-01       2.00        NaN        NaN  
            2010-03-01       0.80        NaN        NaN  
            2010-04-01       2.50       2.00        NaN  
            2010-05-01       3.34       0.80        NaN  
            2010-06-01       6.63       2.50        NaN  
            2010-08-01       9.96       3.34        NaN  
            2010-09-01       5.50       6.63       2.00  
            2010-10-01       4.76       9.96       0.80  
            2010-11-01       1.67       5.50       2.50  
            2010-12-01       1.55       4.76       3.34  
            2

In [9]:
data.dropna(subset=['PRCP'] + [f'PRCP_Lag{lag}' for lag in [1, 3, 7]], inplace=True)

In [10]:
data.head(20)

NAME  LATITUDE  LONGITUDE  \
STATION     DATE                                                           
USC00114355 2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-07-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2011-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2012-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2012-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2012-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
            2012-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   

                        ELEVATION  AWND  PRCP  SNOW  TAVG  TMAX  TMIN  \
STATION     DATE                                                        
USC00114355 2010-09-01      167.6   NaN  4.76   0.0   NaN  76.6   NaN   
            2010-10-01      167.6   NaN  1.67   0.0   NaN  69.9   NaN   
            2010-11-01      167.6   NaN  1.55   0.0   NaN  52.8   NaN   
            2010-12-01      167.6   NaN  1.55  11.0   NaN  30.4   NaN   
            2011-01-01      167.6   NaN  0.89   NaN   NaN  26.6   NaN   
            2011-02-01      167.6   NaN  2.67   NaN   NaN  35.4   NaN   
            2011-03-01      167.6   NaN  2.07   NaN   NaN  48.4   NaN   
            2011-04-01      167.6   NaN  3.64   NaN   NaN  58.7   NaN   
            2011-05-01      167.6   NaN  3.51   NaN   NaN  73.4   NaN   
            2011-06-01      167.6   NaN  4.71   NaN  71.7  80.9  62.6   
            2011-07-01      167.6   NaN  2.02   0.0  80.6  89.8  71.4   
            2011-08-01      167.6   NaN  0.81   NaN  75.1  85.1    65   
            2011-09-01      167.6   NaN  2.21   NaN  62.3  70.9  53.7   
            2011-10-01      167.6   NaN  0.87   NaN    56  66.9  45.2   
            2011-11-01      167.6   NaN  2.59   NaN  43.1  51.7  34.6   
            2011-12-01      167.6   NaN  1.87   0.0  33.7    41  26.3   
            2012-01-01      167.6   NaN  0.62   6.9  28.7  37.9  19.5   
            2012-02-01      167.6   NaN  1.86   NaN  32.1  39.6  24.5   
            2012-03-01      167.6   NaN  1.50   NaN  52.8  63.3  42.3   
            2012-04-01      167.6   NaN  4.50   0.0  53.1  64.1  42.2   

                        PRCP_Lag1  PRCP_Lag3  PRCP_Lag7  
STATION     DATE                                         
USC00114355 2010-09-01       5.50       6.63       2.00  
            2010-10-01       4.76       9.96       0.80  
            2010-11-01       1.67       5.50       2.50  
            2010-12-01       1.55       4.76       3.34  
            2011-01-01       1.55       1.67       6.63  
            2011-02-01       0.89       1.55       9.96  
            2011-03-01       2.67       1.55       5.50  
            2011-04-01       2.07       0.89       4.76  
            2011-05-01       3.64       2.67       1.67  
            2011-06-01       3.51       2.07       1.55  
            2011-07-01       4.71       3.64       1.55  
            2

In [11]:
data.reset_index(inplace=True)

In [12]:
data.head()

STATION       DATE                         NAME  LATITUDE  LONGITUDE  \
0  USC00114355 2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
1  USC00114355 2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
2  USC00114355 2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
3  USC00114355 2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
4  USC00114355 2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   

   ELEVATION  AWND  PRCP  SNOW TAVG  TMAX TMIN  PRCP_Lag1  PRCP_Lag3  \
0      167.6   NaN  4.76   0.0  NaN  76.6  NaN       5.50       6.63   
1      167.6   NaN  1.67   0.0  NaN  69.9  NaN       4.76       9.96   
2      167.6   NaN  1.55   0.0  NaN  52.8  NaN       1.67       5.50   
3      167.6   NaN  1.55  11.0  NaN  30.4  NaN       1.55       4.76   
4      167.6   NaN  0.89   NaN  NaN  26.6  NaN       1.55       1.67   

   PRCP_Lag7  
0       2.00  
1       0.80  
2       2.50  
3       3.34  
4       6.63

In [13]:
# Check for missing values
data.isnull().sum()

STATION         0
DATE            0
NAME            0
LATITUDE        0
LONGITUDE       0
ELEVATION       0
AWND          649
PRCP            0
SNOW         1628
TAVG           16
TMAX            6
TMIN           14
PRCP_Lag1       0
PRCP_Lag3       0
PRCP_Lag7       0
dtype: int64

In [14]:
# Drop the 'SNOW' column
data_cleaned = data.drop(['SNOW'], axis=1)

In [15]:
# Convert non-numeric to numeric
for column in ['AWND', 'TAVG', 'TMAX', 'TMIN']:
    data_cleaned[column] = pd.to_numeric(data_cleaned[column], errors='coerce')

In [16]:
imputer = SimpleImputer(strategy='median')
data_cleaned[['AWND', 'TAVG', 'TMAX', 'TMIN']] = imputer.fit_transform(data_cleaned[['AWND', 'TAVG', 'TMAX', 'TMIN']])

In [17]:
data_cleaned.head(20)

STATION       DATE                         NAME  LATITUDE  LONGITUDE  \
0   USC00114355 2010-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
1   USC00114355 2010-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
2   USC00114355 2010-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
3   USC00114355 2010-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
4   USC00114355 2011-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
5   USC00114355 2011-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
6   USC00114355 2011-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
7   USC00114355 2011-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
8   USC00114355 2011-05-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
9   USC00114355 2011-06-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
10  USC00114355 2011-07-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
11  USC00114355 2011-08-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
12  USC00114355 2011-09-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
13  USC00114355 2011-10-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
14  USC00114355 2011-11-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
15  USC00114355 2011-12-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
16  USC00114355 2012-01-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
17  USC00114355 2012-02-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
18  USC00114355 2012-03-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   
19  USC00114355 2012-04-01  ILLINOIS CITY DAM 16, IL US   41.4255   -91.0094   

    ELEVATION  AWND  PRCP  TAVG  TMAX  TMIN  PRCP_Lag1  PRCP_Lag3  PRCP_Lag7  
0       167.6   7.8  4.76  57.5  76.6  47.6       5.50       6.63       2.00  
1       167.6   7.8  1.67  57.5  69.9  47.6       4.76       9.96       0.80  
2       167.6   7.8  1.55  57.5  52.8  47.6       1.67       5.50       2.50  
3       167.6   7.8  1.55  57.5  30.4  47.6       1.55       4.76       3.34  
4       167.6   7.8  0.89  57.5  26.6  47.6       1.55       1.67       6.63  
5       167.6   7.8  2.67  57.5  35.4  47.6       0.89       1.55       9.96  
6       167.6   7.8  2.07  57.5  48.4  47.6       2.67       1.55       5.50  
7       167.6   7.8  3.64  57.5  58.7  47.6       2.07       0.89       4.76  
8       167.6   7.8  3.51  57.5  73.4  47.6       3.64       2.67       1.67  
9       167.6   7.8  4.71  71.7  80.9  62.6       3.51       2.07       1.55  
10      167.6   7.8  2.02  80.6  89.8  71.4       4.71       3.64       1.55  
11      167.6   7.8  0.81  75.1  85.1  65.0       2.02       3.51       0.89  
12      167.6   7.8  2.21  62.3  70.9  53.7       0.81       4.71       2.67  
13      167.6   7.8  0.87  56.0  66.9  45.2       2.21       2.02       2.07  
14      167.6   7.8  2.59  43.1  51.7  34.6       0.87       0.81       3.64  
15      167.6   7.8  1.87  33.7  41.0  26.3       2.59       2.21       3.51  
16      167.6   7.8  0.62  28.7  37.9  19.5       1.87       0.87       4.71  
17      167.6   7.8  1.86  32.1  39.6  24.5       0.62       2.59       2.02  
18      167.6   7.8  1.50  52.8  63.3  42.3       1.86       1.87       0.81  
19      167.6   7.8  4.50  53.1  64.1  42.2       1.50       0.62       2.21

## Data Partition

In [18]:
data_cleaned['YEAR'] = data_cleaned['DATE'].dt.year
train_data = data_cleaned[data_cleaned['YEAR'] < data_cleaned['YEAR'].max()]
test_data = data_cleaned[data_cleaned['YEAR'] == data_cleaned['YEAR'].max()]

In [19]:
X_train = train_data[[f'PRCP_Lag{lag}' for lag in [1, 3, 7]]]
X_train.head()

PRCP_Lag1  PRCP_Lag3  PRCP_Lag7
0       5.50       6.63       2.00
1       4.76       9.96       0.80
2       1.67       5.50       2.50
3       1.55       4.76       3.34
4       1.55       1.67       6.63

In [20]:
y_train = train_data['PRCP']
X_test = test_data[[f'PRCP_Lag{lag}' for lag in [1, 3, 7]]]
y_test = test_data['PRCP']

In [22]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import mean_squared_error

In [23]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test) 

In [24]:
y_train.isnull().any().any()

False

In [25]:
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X_train_scaled, y_train)
y_test_pred = knn.predict(X_test_scaled)
y_test_pred[:5]
y_train_pred = knn.predict(X_train_scaled)

In [26]:
mse_train = mean_squared_error(y_train, y_train_pred)
mse_train

2.845744840731071

In [27]:
mse_test = mean_squared_error(y_test, y_test_pred)
mse_test 

2.1179003333333335

The average squared difference between the predicted and actual precipitation values of the model is 2.1179.